### PR TITLE
refactor(gcli): download examples from gear-dapp

### LINKED ISSUE #2750: Fetch examples from gear-dapps in gcli
> Really great step forward to get rid of obsolete tests. I hope we will omit all of them soon as we still can't run `cargo install wasm-proc --git https://github.com/gear-tech/gear.git`
> 
> @clearloop Please consider this change, as `gcli` uses examples as a template for a new smart contract.

hmmm, I think after this, I can do https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28#list-organization-repositories for https://github.com/orgs/gear-dapps/

_Originally posted by @clearloop in https://github.com/gear-tech/gear/issues/2749#issuecomment-1587103881_

### PR DESCRIPTION
**Pull request by clearloop**
Resolves #2750 

- [x] download example dapp from `gear-dapps`
- [x] download example to specified path

**short version**

```
 7:53PM ../target/debug/gcli new 
Cloning into 'app'...
remote: Enumerating objects: 32, done.
remote: Counting objects: 100% (32/32), done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 32 (delta 2), reused 14 (delta 0), pack-reused 0
Receiving objects: 100% (32/32), 37.45 KiB | 445.00 KiB/s, done.
Resolving deltas: 100% (2/2), done.
Successfully created app!
```

**invalid app**

```
../target/debug/gcli new --template invalid-app 
Available templates: [
    "dao-app",
    "dao",
    "fungible-token",
    "dao-light",
    "ping",
    "escrow",
    "game-of-chance",
    "feeds",
    "multitoken",
    "non-fungible-token",
    "app",
    "gear-lib",
    "concert",
    "dutch-auction",
    "docs",
    "supply-chain",
    "rock-paper-scissors",
    "RMRK",
    "staking",
    "multisig-wallet",
    "crowdsale",
    "gear-smart-contract-academy",
    "dex",
    "nft-pixelboard",
    "sharded-fungible-token",
    "nft-marketplace",
    "on-chain-nft",
    "syndote",
    "DeIn",
    "sharded-multitoken",
]
```

**long version**
```
../target/debug/gcli new --template app my-app  
Cloning into 'my-app'...
remote: Enumerating objects: 32, done.
remote: Counting objects: 100% (32/32), done.
remote: Compressing objects: 100% (29/29), done.
remote: Total 32 (delta 2), reused 14 (delta 0), pack-reused 0
Receiving objects: 100% (32/32), 37.45 KiB | 479.00 KiB/s, done.
Resolving deltas: 100% (2/2), done.
Successfully created app!
```

@gear-tech/ecosystem-devs

### CODE REVIEW

**Review by shamilsan — APPROVED** [review]
Great 👍 
I think using `gear-dapps` is more appropriate indeed, as they are more advanced and relevant.

File: gcli/src/cmd/new.rs [hunk 1]
```diff
@@ -17,30 +17,30 @@
...
+        let templates = template::list().await?;
 
-            println!("Successfully created {template}!");
+        let template = &self.template;
+        if templates.contains(template) {
+            let path = self.path.as_deref().unwrap_or(template);
+            template::download(template, path).await?;
+            println!("Successfully created {path}!");
         } else {
             println!("Available templates: {:#?}", templates);
```
Comment (breathx) [inline gcli/src/cmd/new.rs:45]: I'd sort alphabetically templates + bold or underline default one

Comment (breathx) [inline gcli/src/cmd/new.rs:45]: Say that given arg is invalid

File: gcli/src/template.rs [hunk 2]
```diff
@@ -18,157 +18,58 @@
...
-            let t = template?.file_name();
-            let example = t.to_string_lossy();
-            if example.contains('.') {
-                continue;
-            }
+/// Download example
+pub async fn download(example: &str, path: &str) -> Result<()> {
+    let url = format!("{}{}.git", GEAR_DAPP_ORG, example);
+    Command::new("git")
+        .args(["clone", &url, path, "--depth=1"])
```
[the remark this answers is not in the document]
Comment (breathx) [reply gcli/src/template.rs:62]: > users download these examples for their own projects but not for contributing to the examples, so I think we even can remove the `.git` and create a new one for them tbh

yep, sounds nice
btw don't forget then adapt toml anyway is it was before (author etc)

File: .github/workflows/CI.yaml [hunk 3]
```diff
@@ -62,7 +62,7 @@ jobs:
       - name: "Install: Build deps"
         run: |
           sudo apt update
-          sudo apt install -y git clang curl libssl-dev llvm libudev-dev cmake protobuf-compiler wget bzip2
+          sudo apt install -y git clang curl libssl-dev llvm libudev-dev cmake protobuf-compiler wget bzip2 pkg-config
```
[the remark this answers is not in the document]
Comment (breathx) [reply .github/workflows/CI.yaml:65]: then we force users to have this pkg-config
better to add the check in cli command by asserting e.g. `which pkg-config` result

File: gcli/src/cmd/new.rs (hunk 1 above)
Comment (breathx) [reply gcli/src/cmd/new.rs]: > https://docs.rs/dialoguer/latest/dialoguer/
> 
> actually we can go with this library to enhance these stuffs in the future, for example, sth like
> 
> Screenshot 2023-06-13 at 15 24 23

wow, that's cool
to be done in future, please open an issue for that

File: .github/workflows/CI.yaml (hunk 3 above)
[the remark this answers is not in the document]
Comment (breathx) [reply .github/workflows/CI.yaml:65]: that won't work for case when you download gcli binary from somewhere, so I'd add the check inside command itself

[the remark this answers is not in the document]
Comment (breathx) [reply .github/workflows/CI.yaml:65]: I suppose we will publish gcli someday, e.g., `brew install gcli`. Furthermore, not only linux will require the dep, but any unix, isn't it?

**Pull request merged**